graph-builder/src/registry.rs

```diff
@@ -37,23 +45,109 @@ impl Into<cincinnati::Release> for Release {
     }
 }
 
+type RegistryClient = dkregistry::v2::Client;
+
+fn registry_trim(src: &str) -> &str {
+    src.trim_left_matches("https://")
+        .trim_left_matches("http://")
+}
+
+fn get_credentials_or_empty_reader(
+    credentials_file: &str,
+) -> Result<Box<dyn std::io::Read>, Error> {
+    if credentials_file != "" {
+        match File::open(credentials_file) {
+            Ok(file) => Ok(Box::new(file)),
+            Err(err) => bail!(err),
+        }
+    } else {
+        Ok(Box::new(std::io::empty()))
+    }
+}
+
+fn get_authenticated_client(
+    tcore: &mut Core,
+    host: &str,
+    image: &str,
+    credentials_file: &str,
+) -> Result<RegistryClient, Error> {
+    let mut registry_client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(host)
+        .insecure_registry(false)
+        .read_credentials(get_credentials_or_empty_reader(credentials_file)?)
+        .build()
+        .map_err(|e| format_err!("{}", e))?;
+
+    let fut_check_v2_prelogin = registry_client
+        .is_v2_supported()
+        .map_err(|e| format_err!("{}", e))?;
+    if !tcore
+        .run(fut_check_v2_prelogin)
+        .map_err(|e| format_err!("{}", e))?
+    {
+        return Err(format_err!("API v2 not supported by {}", host));
+    };
+
+    let fut_check_auth_prelogin = registry_client
+        .is_auth(None)
+        .map_err(|e| format_err!("{}", e))?;
+    let logged_in = tcore
+        .run(fut_check_auth_prelogin)
+        .map_err(|e| format_err!("{}", e))?;
+    if logged_in {
+        return Err(format_err!("no login performed, but already authenticated"));
+    }
+
+    let fut_login = registry_client
+        .login(vec![&format!("repository:{}:pull", image)])
+        .map_err(|e| format_err!("{}", e))?;
+    let token = tcore.run(fut_login).map_err(|e| format_err!("{}", e))?;
+
+    let fut_auth = registry_client
+        .is_auth(Some(token.token()))
+        .map_err(|e| format_err!("{}", e))?;
+    if !tcore.run(fut_auth).map_err(|e| format_err!("{}", e))? {
+        return Err(format_err!("login failed"));
+    }
+    registry_client.set_token(Some(token.token()));
+    debug!("successfully logged in to {}", host);
+
+    let fut_check_v2_postlogin = registry_client
+        .is_v2_supported()
+        .map_err(|e| format_err!("{}", e))?;
+    if !tcore
```

Comment:
Instead of evaluating each of these futures independently, just chain them together and invoke them once.